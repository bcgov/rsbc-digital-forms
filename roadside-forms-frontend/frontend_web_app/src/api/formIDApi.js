import { api } from "./config/axiosConfig";
import { createRequestHeader } from "../utils/requestHeaders";

export const FormIDApi = {
  post: async function (data) {
    const headers = {
      ...await createRequestHeader(),
    };

    try {
      const response = await api
        .request({
          url: "/api/v1/forms",
          method: "POST",
          headers: { ...headers },
          data: JSON.stringify(data),
        });
      return response.data;      
    } catch (error) {
      console.log(error);
    }
    return null;
  },

  get: async function () {
    const headers = {
      ...await createRequestHeader(),
    };
    return await api
      .request({
        url: "/api/v1/forms",
        method: "GET",
        headers: { ...headers },
      })
      .then((response) => {
        return response.data;
      })
      .catch((error) => {
        return {
          status: error.status,
          data: error.response,
        };
      });
  },
  patch: async function (data) {
    const headers = {
      ...await createRequestHeader(),
    };
    return await api
      .request({
        url: "/api/v1/forms",
        method: "PATCH",
        headers: { ...headers },
        data: { ...data },
      })
      .then((response) => {
        return response.data;
      })
      .catch((error) => {
        return {
          status: error.status,
          data: error.response,
        };
      });
  },
};
