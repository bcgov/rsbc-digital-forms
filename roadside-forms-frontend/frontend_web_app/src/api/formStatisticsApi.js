import { api } from "./config/axiosConfig";
import { createRequestHeader } from "../utils/requestHeaders";

export const FormStatisticsApi = {
  getStatistics: async function () {
    const headers = {
      ...await createRequestHeader(),
    };
    return await api
      .request({
        url: "/api/v1/forms/statistics", // Adjust this to match your API endpoint
        method: "GET",
        headers: { ...headers },
      })
      .then((response) => {
        return {
          status: response.status,
          data: response.data,
        };
      })
      .catch((error) => {
        return {
          status: error.status,
          data: error.response,
        };
      });
  },
};