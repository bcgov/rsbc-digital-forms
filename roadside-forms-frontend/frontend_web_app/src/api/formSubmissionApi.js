import { api } from "./config/axiosConfig";
import { createRequestHeader } from "../utils/requestHeaders";

export const FormSubmissionApi = {

    post: async function (data) {
      const headers = {
        ...createRequestHeader(),
      };
      console.log(data)
        const response = await api.request({
          url: "/api/v1/event",
          method: "POST",
          headers: {...headers},
          data: {...data},
        })
        console.log(response)
        return response.json()
      },
};