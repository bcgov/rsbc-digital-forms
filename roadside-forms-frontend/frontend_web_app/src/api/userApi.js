import { api } from "./config/axiosConfig";
import { createRequestHeader } from "../utils/requestHeaders";

export const UserApi = {

    getAll: async function () {
        const headers = {
          ...createRequestHeader(),
        };
        return await api.request({
          url: "/api/v1/admin/users",
          method: "GET",
          headers:{...headers},
        }).then ( (response) => {
          return {
              status: response.status,
              data: response.data
          }
      }).catch((error) =>{
          return {
              status: error.status,
              data: error.response
          }
      })
      },
      delete: async function (data) {
        const headers = createRequestHeader();
          return await api.request({
            url: `/api/v1/admin/users/${data.user_guid}/roles/${data.role_name}`,
            method: "DELETE",
            headers:{...headers},
          }).then ( (response) => {
            return {
                status: response.status,
                data: response.data
            }
        }).catch((error) =>{
            // console.log(error)
            return {
                status: error.status,
                data: error.response
            }
        })
    
        },
      get: async function (userId) {
        const headers = {
          ...createRequestHeader(),
        };
        return await api.request({
          url: `/api/v1/users/${userId}`,
          method: "GET",
          headers:{...headers},
        }).then ( (response) => {
          return {
              status: response.status,
              data: response.data
          }
      }).catch((error) =>{
          // console.log(error)
          return {
              status: error.status,
              data: error.response
          }
      })
      },
    post: async function (data) {
      const headers = createRequestHeader();
        return await api.request({
          url: "/api/v1/users",
          method: "POST",
          headers:{...headers},
          data: {...data},
        }).then ( (response) => {
          return {
              status: response.status,
              data: response.data
          }
      }).catch((error) =>{
          // console.log(error)
          return {
              status: error.status,
              data: error.response
          }
      })
  
      },
      patch: async function (data) {
        const headers = createRequestHeader();
          return await api.request({
            url: `/api/v1/admin/users/${data.user_guid}/roles/officer`,
            method: "PATCH",
            headers:{...headers},
          }).then ( (response) => {
            return {
                status: response.status,
                data: response.data
            }
        }).catch((error) =>{
            // console.log(error)
            return {
                status: error.status,
                data: error.response
            }
        })
        },
        postAdmin: async function (data) {
          const headers = createRequestHeader();
            return await api.request({
              url: `/api/v1/admin/users/${data.user_guid}/roles`,
              method: "POST",
              data: {"role_name": "administrator"},
              headers:{...headers},
            }).then ( (response) => {
              return {
                  status: response.status,
                  data: response.data
              }
          }).catch((error) =>{
              // console.log(error)
              return {
                  status: error.status,
                  data: error.response
              }
          })
          },
};
