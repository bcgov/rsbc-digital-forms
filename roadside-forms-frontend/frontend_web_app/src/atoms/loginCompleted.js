import {atom} from 'recoil'

export const loginCompletedAtom = atom({
    key: 'loginCompleted',
    default: false,
  });