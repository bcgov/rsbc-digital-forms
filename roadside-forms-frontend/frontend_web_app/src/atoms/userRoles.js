import {atom} from 'recoil'

const userRolesAtom = atom({
    key: 'userRoles',
    default: [],
  });