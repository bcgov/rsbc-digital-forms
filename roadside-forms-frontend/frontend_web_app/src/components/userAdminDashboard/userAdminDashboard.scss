.delete{
    background-color: #ff0000;
}

.approve{
    background-color: #008000;
}

.flex{
    display: flex;
}

.right{
    text-align: right;
}

.top-margin{
 margin-top: 32px;
}