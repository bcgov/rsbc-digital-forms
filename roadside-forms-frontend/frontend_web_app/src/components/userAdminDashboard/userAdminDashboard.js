import React, {useState,useEffect} from 'react';
import { Formik, Form} from 'formik';
import { useNavigate} from 'react-router-dom';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import moment from 'moment-timezone';
import { useRecoilValue } from 'recoil';
import { userRolesAtom } from '../../atoms/userRoles';


import { UserApi } from '../../api/userApi';
import { SearchableSelect } from '../common/Select/SearchableSelect';
import './userAdminDashboard.scss';

export const UserAdminDashboard = () => {

    const initialValues = {
      user:{},
    };

    const [isLoading, setLoading] = useState(true)
    const [data, setData] = useState({});
    const [selectUsers, setSelectUsers] = useState([]);
    const userRoleData = useRecoilValue(userRolesAtom)
    const navigate = useNavigate();
    
    useEffect(() => {
      if(!userRoleData.some( role => role['role_name'] === 'administrator' )){
        navigate("/")
      }else{
      getAllUsers();
      }
    }, [userRoleData,navigate]);
    
    const getAllUsers = () =>{
      UserApi.getAll().then( (resp) => {
        setSelectUsers(resp.data.filter(user => user.role_name === "officer" ).map((user) => { return ({"label":user.login,"value":user})}));
        setData(resp.data);
        setLoading(false);
      })
    }

    const deleteUser = (user) => {
      UserApi.delete(user).then(()=>{
        setLoading(true);
        getAllUsers();
      })
    }
    
    const approveUser = (user) => {
      UserApi.patch(user).then(()=>{
        setLoading(true);
        getAllUsers();
      })
    }

    const onSubmit = (values, { setSubmitting }) => {
      UserApi.postAdmin(values.user).then(() => {
        setSubmitting(false);
        getAllUsers();
      })
      
    }
  
    if(isLoading){
      return <div className="App">Loading...</div>;
    }else{
      return (
        <>
          <div className='border-design text-font'>
          <Table>
            <thead>
              <tr>
                <th>SURNAME</th>
                <th>Given</th>
                <th>Prime ID</th>
                <th>Agency</th>
                <th>Keycloak GUID</th>
                <th>Username</th>
                <th>Role</th>
                <th>Date Applied</th>
                <th>Action</th>
              </tr>
            </thead>
            <tbody>
              {data && data.map((user) => (
                <tr key={`${user.user_guid}-${user.role_name}`}>
                  <td>{user.last_name}</td>
                  <td>{user.first_name}</td>
                  <td>{user.badge_number}</td>
                  <td>{user.agency}</td>
                  <td>{user.user_guid}</td>
                  <td>{user.login}</td>
                  <td>{user.role_name}</td>
                  <td className="text-muted small">{moment(user.submitted_dt).tz("America/Vancouver").format("YYYY-MM-DD HH:mm")}</td>
                  <td>{user.approved_dt ? <Button variant="danger" onClick={() => {deleteUser(user)}} >Delete</Button> : <Button variant="success" onClick={() => {approveUser(user)}} >Approve</Button>}</td>
                </tr>
              ))}
            </tbody>
          </Table>
          <Formik initialValues={initialValues} onSubmit={onSubmit}>
          {({ isSubmitting }) => (
            <Form>
              <Container fluid>
                <Row>
                  <Col sm={10}>
                    <SearchableSelect label="Grant administrator role to selected user" className="field-width field-height" name="user" options={selectUsers} />
                  </Col>
                  <Col sm={2} className="top-margin">
                    <Button type="submit" variant="primary" size="sm" disabled={isSubmitting}>Add as administrator</Button>
                  </Col>
                </Row>
              </Container>
            </Form>
          )}
          </Formik>
        </div>
      </>
      );
    }
  }
