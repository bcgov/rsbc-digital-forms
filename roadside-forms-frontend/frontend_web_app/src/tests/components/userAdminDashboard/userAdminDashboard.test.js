import React from 'react';
import { render, act } from '@testing-library/react';
import { RecoilRoot } from 'recoil';
import { userRolesAtom } from '../../../atoms/userRoles'; 
import { loginCompletedAtom } from '../../../atoms/loginCompleted'; 
import { useNavigate } from 'react-router-dom';
import { UserApi } from '../../../api/userApi';
import { UserAdminDashboard } from '../../../components/userAdminDashboard/userAdminDashboard';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('UserAdminDashboard component', () => {
  test('should match snapshot', async () => {
    const mockNavigate = jest.fn();

    UserApi.getAll = jest.fn().mockResolvedValue({
      data: [
        {
          agency: "BCHP Burnaby",
          approved_dt: "Thu, 22 Jun 2023 16:51:39 GMT",
          badge_number: "222222",
          display_name: "Test, user",
          first_name: "USER",
          last_name: "Test",
          login: "TEST@idir",
          role_name: "administrator",
          submitted_dt: "Thu, 22 Jun 2023 16:51:39 GMT",
          user_guid: "91D388D1339C41388E622F5",
          username: "91d388d1339c41388e622f5@idir",
        },
        {
          agency: "BCHP Burnaby",
          approved_dt: "Thu, 22 Jun 2023 16:51:39 GMT",
          badge_number: "222222",
          display_name: "Test, user",
          first_name: "USER",
          last_name: "Test",
          login: "TEST@idir",
          role_name: "officer",
          submitted_dt: "Thu, 22 Jun 2023 16:51:39 GMT",
          user_guid: "91D388D1339C41388E622F5",
          username: "91d388d1339c41388e622f5@idir",
        }
      ]
    });

    const setSelectUsersMock = jest.fn();
    const setDataMock = jest.fn();
    const setLoadingMock = jest.fn();

    useNavigate.mockImplementation(() => mockNavigate);

    let container; 

    await act(async () => {
      ({ container } = render( 
        <RecoilRoot
          initializeState={(snap) => {
            snap.set(userRolesAtom, [
              {
                approved_dt: "Mon, 05 Jun 2023 14:27:31 GMT",
                role_name: "administrator",
                submitted_dt: "Thu, 01 Jun 2023 09:39:42 GMT",
                user_guid: "91d388d1889c87654e622f5e9@idir"
              }
            ]);
            snap.set(loginCompletedAtom, true);
          }}
        >
          <UserAdminDashboard setLoading={setLoadingMock} setSelectUsers={setSelectUsersMock} setData={setDataMock} />
        </RecoilRoot>
      ));
    });

    expect(container).toMatchSnapshot();
  });
});
