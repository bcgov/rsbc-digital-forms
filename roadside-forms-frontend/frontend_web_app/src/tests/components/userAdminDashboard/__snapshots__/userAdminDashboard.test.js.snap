// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAdminDashboard component should match snapshot 1`] = `
<div>
  <div
    class="border-design text-font"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            SURNAME
          </th>
          <th>
            Given
          </th>
          <th>
            Prime ID
          </th>
          <th>
            Agency
          </th>
          <th>
            Keycloak GUID
          </th>
          <th>
            Username
          </th>
          <th>
            Role
          </th>
          <th>
            Date Applied
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Test
          </td>
          <td>
            USER
          </td>
          <td>
            222222
          </td>
          <td>
            BCHP Burnaby
          </td>
          <td>
            91D388D1339C41388E622F5
          </td>
          <td>
            TEST@idir
          </td>
          <td>
            administrator
          </td>
          <td
            class="text-muted small"
          >
            2023-06-22 09:51
          </td>
          <td>
            <button
              class="btn btn-danger"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Test
          </td>
          <td>
            USER
          </td>
          <td>
            222222
          </td>
          <td>
            BCHP Burnaby
          </td>
          <td>
            91D388D1339C41388E622F5
          </td>
          <td>
            TEST@idir
          </td>
          <td>
            officer
          </td>
          <td
            class="text-muted small"
          >
            2023-06-22 09:51
          </td>
          <td>
            <button
              class="btn btn-danger"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <form
      action="#"
    >
      <div
        class="container-fluid"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-10"
          >
            <div>
              <label
                for="user"
              >
                Grant administrator role to selected user
              </label>
              <div
                class="field-width field-height css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class=" css-13cymwt-control"
                >
                  <div
                    class=" css-1fdsijx-ValueContainer"
                  >
                    <div
                      class=" css-1dimb5e-singleValue"
                    />
                    <div
                      class=" css-qbdosj-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-haspopup="true"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="user-select"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-1u9des2-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <input
                  name="user"
                  type="hidden"
                  value=""
                />
              </div>
              <input
                id="user"
                name="user"
                type="hidden"
                value="{}"
              />
            </div>
          </div>
          <div
            class="top-margin col-sm-2"
          >
            <button
              class="btn btn-primary btn-sm"
              type="submit"
            >
              Add as administrator
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
